runtime/bcverify/vrfyhelp.c

```diff
@@ -417,6 +417,15 @@ pushLdcType(J9ROMClass * romClass, UDATA index, UDATA * stackTop)
 		case J9CPTYPE_METHODHANDLE:
 			PUSH(BCV_JAVA_LANG_INVOKE_METHODHANDLE_INDEX << BCV_CLASS_INDEX_SHIFT);
 			break;
+		case J9CPTYPE_CONSTANT_DYNAMIC:
+			J9ROMConstantDynamicRef* romConstantDynamicRef = (J9ROMConstantDynamicRef *)(J9_ROM_CP_FROM_ROM_CLASS(romClass) + index);
+			J9UTF8 *nameAndSignature = J9ROMCONSTANTDYNAMICREF_NAMEANDSIGNATURE(romConstantDynamicRef);
+			U_8* signature = J9UTF8_DATA(nameAndSignature);
+
+			/* The return value of a method follows the first ')' listed in its signature */
+			while (*signature++ != ')');
```

Comment:
read the condy spec 4.4.13, it seem to suggest that the signature referenced in a ConstantDynamic entry is a fieldDescriptor, so there will be no `()` in the string, it should only contain the return type.

I'll read the spec details to confirm this, may need to add more checks in static verify to consider cases such as double array signature `[D`

Reply:
Right, thanks. Looks like I could directly give the value of `J9UTF8_DATA(nameAndSignature)` then, and `pushType` will still handle the rest. 

What situations istatic verify would/would not accept cases such as the double array signature `[[D`? From what I understand:
- Reference types are any non-primitive (eg. objects, arrays)
- Constant dynamic does not place any restrictions on its return value (4.10.1.3)
> dynamic(ConstantName, FieldDescriptor) for a CONSTANT_Dynamic_info, where ConstantName and FieldDescriptor correspond to the name and field descriptor referenced by the name_and_type_index item of the CONSTANT_Dynamic_info structure (the bootstrap_method_attr_index item is irrelevant to verification);

&

> loadableConstantType(CP, Type) :-
CP = dynamic(ConstantName, FieldDescriptor),
parseFieldDescriptor(FieldDescriptor, Type).
- `ldc` & `ldc_w` bytecodes  would still accept the reference types (including double array) and primitives which are not long/double
- `ldc2_w` bytecode would still only accept long/double - but not rerfences, such as `[[D`



---

Quick reread of the section you pointed out gives:

4.4.13 gives:
> 4.4.13 The CONSTANT_Dynamic_info Structure
>  CONSTANT_Dynamic_info {
&nbsp;&nbsp;&nbsp; u1 tag;
&nbsp;&nbsp;   u2 bootstrap_method_attr_index;
&nbsp;&nbsp;   u2 name_and_type_index;
}
...
The value of the name_and_type_index item must be a valid index into the constant_pool table. The constant_pool entry at that index must be a CONSTANT_NameAndType_info structure (4.4.6) representing a name and field descriptor (4.3.2).

4.4.6 gives:
> CONSTANT_NameAndType_info {
&nbsp;&nbsp;&nbsp;    u1 tag;
&nbsp;&nbsp;&nbsp;    u2 name_index;
&nbsp;&nbsp;&nbsp;    u2 descriptor_index;
}
...
The value of the descriptor_index item must be a valid index into the constant_pool table. The constant_pool entry at that index must be a CONSTANT_Utf8_info structure (§4.4.7) representing a valid field descriptor or method descriptor (§4.3.2, §4.3.3).

4.3.2 gives the grammar for a field descriptor as:
> FieldDescriptor:
&nbsp;&nbsp;&nbsp;FieldType
FieldType:
&nbsp;&nbsp;&nbsp;BaseType   |  ObjectType  |  ArrayType
BaseType:
&nbsp;&nbsp;&nbsp; (one of) B C D F I J S Z
ObjectType:
&nbsp;&nbsp;&nbsp;L ClassName ;
ArrayType:
&nbsp;&nbsp;&nbsp;[ ComponentType
ComponentType:
&nbsp;&nbsp;&nbsp;FieldType
(http://cr.openjdk.java.net/~dlsmith/constant-dynamic.html)

Reply:
Woops, you're right - sounds like I can put directly into `pushType`

I'm not seeing any restrictions on the type of the signature - references (including cases like `[[D`) are valid to be passed to `ldc` or `ldc_w`

Reply:
your current check is checking at the last `char` of the fieldDescriptor, which would mix up `[[D` and `D` we would have to either check the first char to be `D` or `J`.
the reverse would apply to `ldc` and `ldc_w`

Reply:
Yes, but if it's a a fieldDescriptor, then can't we directly check the first index instead?
ie. it's either a `BaseType`, `ObjectType`, or `ArrayType`
If a `BaseType`, then one of `B C D F I J S Z` will be the first char; if an `ObjectType`, then `L` will be the first char; and if `ArrayType`, then `[` will be the first char